compiler/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -779,4 +805,129 @@ mod test {
 
         assert_eq!(main.dfg[main.entry_block()].instructions().len(), 3);
     }
+
+    #[test]
+    fn keep_inc_rc_on_borrowed_array_store() {
+        // acir(inline) fn main f0 {
+        //     b0():
+        //       inc_rc [u32 0, u32 0]
+        //       v2 = allocate
+        //       store [u32 0, u32 0] at v2
+        //       jmp b1()
+        //     b1():
+        //       v3 = load v2
+        //       v5 = array_set v3, index u32 0, value u32 1
+        //       return v5
+        //   }
+        let main_id = Id::test_new(0);
+
+        // Compiling main
+        let mut builder = FunctionBuilder::new("main".into(), main_id);
+        let zero = builder.numeric_constant(0u128, Type::unsigned(32));
+        let array_type = Type::Array(Arc::new(vec![Type::unsigned(32)]), 2);
+        let array = builder.array_constant(vector![zero, zero], array_type.clone());
+        builder.increment_array_reference_count(array);
+        let v2 = builder.insert_allocate(array_type.clone());
+        builder.insert_store(v2, array);
+
+        let b1 = builder.insert_block();
+        builder.terminate_with_jmp(b1, vec![]);
+        builder.switch_to_block(b1);
+
+        let v3 = builder.insert_load(v2, array_type);
+        let one = builder.numeric_constant(1u128, Type::unsigned(32));
+        let v5 = builder.insert_array_set(v3, zero, one);
+        builder.terminate_with_return(vec![v5]);
+
+        let ssa = builder.finish();
+        let main = ssa.main();
+
+        // The instruction count never includes the terminator instruction
+        assert_eq!(main.dfg[main.entry_block()].instructions().len(), 3);
+        assert_eq!(main.dfg[b1].instructions().len(), 2);
+
+        // We expect the output to be unchanged
+        let ssa = ssa.dead_instruction_elimination();
+        let main = ssa.main();
+
+        assert_eq!(main.dfg[main.entry_block()].instructions().len(), 3);
+        assert_eq!(main.dfg[b1].instructions().len(), 2);
+    }
+
+    #[test]
+    fn keep_inc_rc_on_borrowed_array_set() {
```

Comment:
Something these tests are missing is a test on how we handle `inc_rc` _after_ when a array is borrowed, e.g.

```
        // acir(inline) fn main f0 {
        //     b0(v0: [u32; 2]):
        //       inc_rc v0
        //       v3 = array_set v0, index u32 0, value u32 1
        //       inc_rc v0
        //       inc_rc v0
        //       inc_rc v0
        //       v4 = array_get v0, index u32 0
        //       return v4
        //   }
```

This would give a little more security against us accidentally removing more instructions than we expect in future changes.

Reply:
done